pkg/reconciler/workload/reconciler.go

```diff
@@ -0,0 +1,179 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workload
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	workloadv1alpha1 "github.com/crossplane/crossplane/apis/workload/v1alpha1"
+)
+
+const (
+	reconcileTimeout = 1 * time.Minute
+	shortWait        = 30 * time.Second
+	longWait         = 1 * time.Minute
+)
+
+// Reconcile error strings.
+const (
+	errGetWorkload          = "cannot get workload"
+	errUpdateWorkloadStatus = "cannot update workload status"
+	errPackageWorkload      = "cannot package workload"
+	errGetPackage           = "cannot get workload package"
+	errApplyWorkloadPackage = "cannot apply workload package"
+)
+
+// Reconcile event reasons.
+const (
+	reasonPackageWorkload = "WorkloadPackaged"
+
+	reasonCannotGetPackage           = "CannotGetWorkloadPackage"
+	reasonCannotPackageWorkload      = "CannotPackageWorkload"
+	reasonCannotApplyWorkloadPackage = "CannotApplyWorkloadPackage"
+)
+
+// A ReconcilerOption configures a Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(l logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = l
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// WithPackager specifies how the Reconciler should package the workload.
+func WithPackager(p Packager) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.packager = p
+	}
+}
+
+// WithApplicator specifies how the Reconciler should apply the workload
+// package.
+func WithApplicator(a resource.Applicator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.applicator = a
+	}
+}
+
+// A Reconciler reconciles an OAM workload type by packaging it into a
+// KubernetesApplication.
+type Reconciler struct {
+	client      client.Client
+	newWorkload func() Workload
+	packager    Packager
+	applicator  resource.Applicator
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Kind is a kind of OAM workload.
+type Kind schema.GroupVersionKind
+
+// A Workload is a type of OAM workload.
+type Workload interface {
+	resource.Conditioned
+	metav1.Object
+	runtime.Object
+}
```

Comment:
It seems like this isn't far off dealing in `*unstructured.Unstructured` - we only really use it to set object metadata and status conditions.